rust/template/lib.rs

```diff
@@ -172,43 +189,81 @@ pub extern "C" fn ddlog_run(
                 }));
                 let prog = program.run(workers as usize);
-                sync::Arc::into_raw(sync::Arc::new(HDDlog{prog: sync::Mutex::new(prog),
-                                                          db: Some(db2),
-                                                          print_err: print_err}))
+                sync::Arc::into_raw(sync::Arc::new(HDDlog{prog:        sync::Mutex::new(prog),
+                                                          db:          Some(db2),
+                                                          print_err:   print_err,
+                                                          replay_file: None}))
             }
         }
     } else {
         match cb {
             None => {
                 let program = prog(sync::Arc::new(|_,_,_| {}));
                 let prog = program.run(workers as usize);
-                sync::Arc::into_raw(sync::Arc::new(HDDlog{prog:sync::Mutex::new(prog),
-                                                          db: None,
-                                                          print_err: print_err}))
+                sync::Arc::into_raw(sync::Arc::new(HDDlog{prog:        sync::Mutex::new(prog),
+                                                          db:          None,
+                                                          print_err:   print_err,
+                                                          replay_file: None}))
             },
             Some(cb) => {
                 let program = prog(sync::Arc::new(move |relid,v,w| {
                         debug_assert!(w == 1 || w == -1);
                         cb(cb_arg, relid, &v.clone().into_record() as *const record::Record, w == 1)
                     }));
                 let prog = program.run(workers as usize);
-                sync::Arc::into_raw(sync::Arc::new(HDDlog{prog: sync::Mutex::new(prog),
-                                                          db: None,
-                                                          print_err: print_err}))
+                sync::Arc::into_raw(sync::Arc::new(HDDlog{prog:        sync::Mutex::new(prog),
+                                                          db:          None,
+                                                          print_err:   print_err,
+                                                          replay_file: None}))
             }
         }
     }
 }
 
+#[no_mangle]
+pub unsafe extern "C" fn ddlog_record_commands(prog: *const HDDlog, fd: unix::io::RawFd) -> raw::c_int
+{
+    if prog.is_null() {
+        return -1;
+    };
+    let mut prog = sync::Arc::from_raw(prog);
+    let res = match sync::Arc::get_mut(&mut prog) {
+        Some(prog) => {
+            /* Swap out the old file and convert it into FD to prevent
+             * the file from closing on destruction (it is the caller's
+             * responsibility to close the file when they're done with it). */
+            let mut old_file = if fd == 0 {
```

Comment:
0 is a legal file descriptor in unix; it's stdout.

Reply:
Good point. Let's use -1 instead.